### PR TITLE
[Pipeline] Add and use RayServeLazyHandle for DAG deployment args

### LINKED ISSUE #23206: [serve] Don't require `serve.start` before `serve.run` [lazy handle]
`serve.run` should work without calling `ray.init` or `serve.start` prior.

```
~/code/ray/python/ray/serve/api.py in _connect()
    761         controller = ray.get_actor(controller_name, namespace=controller_namespace)
    762     except ValueError:
--> 763         raise RayServeException(
    764             "There is no "
    765             "instance running on this Ray cluster. Please "

RayServeException: There is no instance running on this Ray cluster. Please call `serve.start(detached=True) to start one.
```

### PR DESCRIPTION
**Pull request by jiaodong**
## Why are these changes needed?

In current DAG building we replace deployments with corresponding handle in init args, however this requires having `serve.start` running. We don't want to enforce it for local dev or build, thus deferring it to runtime.

closes #23206

## Todo: 
`test_passing_handle` needs to be fixed as it complains `Tried to call a method '__args__' that does not exist. Available methods: [].`

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jiaodong** [comment]
failed tests are due to our use of `getattr` as it's trying to access fields that are not intended to reconfigure the method we access.

(ServeController pid=51594) <<<< getting attribute: __values__
(ServeController pid=51594) <<<< getting attribute: __type__
(ServeController pid=51594) <<<< getting attribute: __type__
(ServeController pid=51594) <<<< getting attribute: __union_params__
(ServeController pid=51594) <<<< getting attribute: __tuple_params__
(ServeController pid=51594) <<<< getting attribute: __result__
(ServeController pid=51594) <<<< getting attribute: __args__
(ServeController pid=51594) <<<< getting attribute: __origin__
(ServeController pid=51594) <<<< getting attribute: __values__
(ServeController pid=51594) <<<< getting attribute: __type__
(ServeController pid=51594) <<<< getting attribute: __type__
(ServeController pid=51594) <<<< getting attribute: __union_params__
(ServeController pid=51594) <<<< getting attribute: __tuple_params__
(ServeController pid=51594) <<<< getting attribute: __result__
(ServeController pid=51594) <<<< getting attribute: __args__

**Comment by simon-mo** [comment]
will rebase once https://github.com/ray-project/ray/pull/23285 is merged

**Comment by jiaodong** [comment]
looking great, btw since the `DeploymentFunctionNode` PR just got merged we need to rebase and use lazy handle by default for that too

**Comment by simon-mo** [comment]
> looking great, btw since the DeploymentFunctionNode PR just got merged we need to rebase and use lazy handle by default for that too

nice catch tests added

File: python/ray/serve/handle.py [hunk 1]
```diff
@@ -234,3 +235,54 @@ def __reduce__(self):
...
         }
         return RayServeSyncHandle._deserialize, (serialized_data,)
+
+
+class RayServeLazySyncHandle:
+    """Lazily initialized handle that only gets fulfilled upon first execution."""
+
+    def __init__(
+        self,
+        deployment_name: EndpointTag,
```
Comment (edoakes) [inline python/ray/serve/handle.py:245]: oh man `EndpointTag` hah

File: python/ray/serve/handle.py [hunk 2]
```diff
@@ -234,3 +235,54 @@ def __reduce__(self):
...
+        handle_options: Optional[HandleOptions] = None,
+    ):
+        self.deployment_name = deployment_name
+        self.handle_options = handle_options or HandleOptions()
+        # For Serve DAG we need placeholder in DAG binding and building without
+        # requirement of serve.start; Thus handle is fulfilled at runtime.
+        self.handle = None
+
+    def options(self, *, method_name: Union[str, DEFAULT] = DEFAULT.VALUE):
```
Comment (edoakes) [inline python/ray/serve/handle.py:254]: looks like the only option we support is `method_name`, do we need to have this generic handling?

File: python/ray/serve/tests/test_pipeline_handle_serde.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+def test_nested_building():
+    with InputNode() as inp:
+        out = func.bind(inp)
+        out = Driver.bind().__call__.bind(out)
+        out = func.bind(out)
+    dag = Driver.bind(out, func.bind())
+    assert len(pipeline_build(dag)) == 5
+
+
+def test_environment_end():
```
Comment (jiaodong) [inline python/ray/serve/tests/test_pipeline_handle_serde.py:54]: nice trick for localized execution

**Review by jiaodong — COMMENTED** [review]
lgtm ! i can't approve it given i created the PR + first a few commits, but you can ;)

**Review by simon-mo — APPROVED** [review]

**Pull request merged**